Code Actions doc page

vzmerr opened the pull request:
It resolves #2296

tgodzik commented on docs/codeactions/code-actions.md:
```diff
@@ -0,0 +1,79 @@
+[Convert To Named Arguments Code Action](https://github.com/scalameta/metals/docs/codeactions/gifs/ConvertToNamedArguments.gif)
+
+
+## Create Companion Object
```
Let's add a short explanation even if it seems obvious

tgodzik commented on docs/codeactions/code-actions.md:
```diff
@@ -0,0 +1,96 @@
+
+Code Actions are there to provide refactoring or rewriting of your current code in order to enhance its readability or save you some time.
+
+The common way of using them, is to click on the yellow light bulb appearing on your code, near the cursor and then, to pick the desired action to peform on the code.
```
```suggestion
In Visual Studio Code, to use them you can click on the yellow light bulb appearing on your code, near the cursor and then, to pick the desired action to perform on the code. Alternatively, you can use the `editor.action.quickFix` shortcut.

For other editors this should also be available via specified shortcut.
```

tgodzik commented on docs/codeactions/code-actions.md:
```diff
@@ -0,0 +1,96 @@
+
+## Create Companion Object
+
+If you have a `trait` or a `class` definition, you can invoke this code action for creating its companion object.
```
```suggestion
If you have a `trait` or a `class` definition, you can invoke this code action to create its companion object.
```

tgodzik commented on docs/codeactions/code-actions.md:
```diff
@@ -0,0 +1,96 @@
+
+## Implement Abstract Members of the Parent Type
+
+Upon inherting from a type, you also have to implement its abstract members. But manuallylooking them all up and copying their signature is time consuming, isn't it? You can just use this code action instead.
```
```suggestion
Upon inherting from a type, you also have to implement its abstract members. But manually looking them all up and copying their signature is time consuming, isn't it? You can just use this code action instead.
```

vzmerr replied:
The informal descriptions are nowadays very popular in technical manuals and blog posts. They entertain and engage the user as they go through the article. So they increase the chance of the article for getting read to the end.

vzmerr replied:
But people are different. However you write it, there will still be people who won't be ok with it, won't there?

tgodzik commented on docs/codeactions/code-actions.md:
```diff
@@ -0,0 +1,96 @@
+
+## Organize Imports
+
+Yes, just have all the clutter, as well as the unnecessary ones removed and tidied up.
```
This is a bit unclear, it would also be good to mention that we need a special compiler option for removing unused imports and that in some editors it's not available on the "lgiht bulb"

tanishiking commented on docs/codeactions/code-actions.md:
```diff
@@ -0,0 +1,92 @@
+
+This code action adds the name to parameters passed to a method call or a class constructor.
+
+[Convert To Named Arguments Code Action](https://github.com/scalameta/metals/docs/codeactions/gifs/ConvertToNamedArguments.gif)
```
```suggestion
![convert to named arguments code action](./gifs/converttonamedarguments.gif)
```

to show the gif in the webpage instead of just a link to gif

tanishiking reviewed (commented):
The document itself looks good, but you need to add `"codeactions/codeactions"` to somewhere in `website/sidebar.json` 👍

vzmerr commented on docs/codeactions/code-actions.md:
```diff
@@ -0,0 +1,92 @@
+### Extract Value
+
+If you think your code would look nicer, if the calcuation of a value is extracted and assined to a new `val`; this code action would come to your aid.
+![Extract Value](./gifs/ExtractValue.gif)
```
ooops, yeah. Thank you so much for taking the time to try this 👍

tanishiking commented on docs/codeactions/code-actions.md:
```diff
@@ -0,0 +1,93 @@
+
+![Rename File](./gifs/RenameFile.gif)
+
+### Extract Value
```
```suggestion
## Extract Value
```

tanishiking commented on docs/codeactions/code-actions.md:
```diff
@@ -0,0 +1,93 @@
+
+![Extract Value](./gifs/ExtractValue.gif)
+
+### map, flatMap, withFilter to For Yield
```
```suggestion
## map, flatMap, withFilter to For Yield
```

tanishiking commented on docs/codeactions/code-actions.md:
```diff
@@ -0,0 +1,93 @@
+
+![Implement Abstract Members](./gifs/ImplementAbstractMembers.gif)
+
+### Import Missing Symbol 
```
```suggestion
## Import Missing Symbol 
```

tanishiking reviewed (approved):
I left nitpicks but LGTM!
BTW, I'm wondering maybe we should have a section (in `sidebar.json`) something like `Guide` and code actions could be under that 🤔 but nevermind, we don't have to work on it in this PR :)

tanishiking commented:
The `yarn build` is failing because

```
These sidebar document ids do not exist:
- codeactions/code-actions
```

instead we should say `codeactions/codeactions`

as the error mentions

```
Available document ids are:
- build-tools/bloop
...
- codeactions/codeactions
```

it doesn't have `hyphen` between `code` and `actions` because you set `id: codeactions` in https://github.com/scalameta/metals/pull/4157/files#diff-0e21d372368917c3f98bf0a8e612f99861c109b89be879f8b1a7bc90fbd9cd35R2

dos65 commented on docs/codeactions/code-actions.md:
```diff
@@ -0,0 +1,93 @@
+
+## Insert Inferred Type
+
+This code action tries to infer the type of variables according to the context, and insert them for you.
```
```suggestion
This code action infers the type of variables according to the context, and insert them for you.
```

dos65 commented on docs/codeactions/code-actions.md:
```diff
@@ -0,0 +1,93 @@
+
+For other editors this should also be available via specified shortcut.
+
+Here is a list of the code actions supported by Metals. Some of these code actions support much more complex use cases, which you would eventually discover through your code, by clicking on the magic 💡, as it appears, to get your surprise!
```
```suggestion
Here is a list of the code actions supported by Metals.
```
I would omit or rephrase the last sentence here :smile_cat:
This part adds some ambiguity: what means `much more complex use cases`, `you would eventually discover` - when, where?

vzmerr replied:
well, the current gifs are only covering the simplest cases for getting the message across. So let's say the use case of covering partial functions in map/flatMap is not shown, so the user would eventually discover this more complex one through clicking the light bulb.

dos65 replied:
> So let's say the use case of covering partial functions in map/flatMap is not shown

From user viewpoint there is no difference. I think, people would be more surprised if it doesn't work in some cases rather than it support some particular tree inside an expression.

> so the user would eventually discover 

Yep, sometimes actions are discovered by a user but I think it's a rare occasion.

I think, in general, interaction with actions comes not from light-bulb suggestion.
User know which actions Metals supports and calls them to avoid manual rewriting.
So, you want to specify a type of your val - you know that there is an action. The same with for-comp conversion and others. Usually you know in advance that some particular code should have a rewrite action.

vzmerr replied:
ok, no problem.

dos65 reviewed (commented):
Nice work! :tada:

tgodzik reviewed (approved):
LGTM! I just moved the gifs to the `static` folder next to other imgs

tgodzik commented:
> LGTM! I just moved the gifs to the `static` folder next to other imgs

Well, that totally failed :| It seems it was working locally, but due to some caching. Reverted the change

vzmerr commented:
> > LGTM! I just moved the gifs to the `static` folder next to other imgs
> 
> Well, that totally failed :| It seems it was working locally, but due to some caching. Reverted the change

perhaps the gifs in the static folder were not added to git, because they do not show up in the corresponding commit
